PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java

```diff
@@ -53,35 +65,77 @@ public Object onExecute(ExecutionEvent event) throws ExecutionException {
+                PublishWizard pw = new PublishWizard();
+                WizardDialog pwd = new PublishWizardDialog(window.getShell(), pw);
+                if (pwd.open() == Window.OK) {
+                    ConsoleLogger.info(String.format("URL: http://%s.azurewebsites.net/",
+                            DockerRuntime.getInstance().getLatestWebAppName()));
+                    ConfigFileUtil.saveConfig(project, DockerRuntime.getInstance().saveToProps(props));
+                }
+            });
+        }, err -> {
             String dockerHost = DockerRuntime.getInstance().getDockerBuilder().uri().toString();
-            String dockerFilePath = project.getFolder(Constant.DOCKER_CONTEXT_FOLDER).getFile(Constant.DOCKERFILE_NAME).getFullPath().toString();
+            String dockerFileRelativePath = Paths.get(basePath).getParent().toUri()
+                    .relativize(Paths.get(basePath, Constant.DOCKERFILE_FOLDER, Constant.DOCKERFILE_NAME).toUri())
+                    .toString();
+            DefaultLoader.getIdeHelper().invokeAndWait(() -> {
+                MessageDialog.openError(window.getShell(), "Error on building image", String
+                        .format(Constant.ERROR_BUILDING_IMAGE, dockerHost, dockerFileRelativePath, err.getMessage()));
+            });
+        });
 
-            MessageDialog.openError(window.getShell(), "Error on building image", String.format(
-                    Constant.ERROR_BUILDING_IMAGE, dockerHost, dockerFilePath.replaceFirst("^/", ""), e.getMessage()));
-            return null;
-        }
-        PublishWizard pw = new PublishWizard();
-        WizardDialog pwd = new PublishWizardDialog(window.getShell(), pw);
-        if (pwd.open() == Window.OK) {
-            ConsoleLogger.info(String.format("URL: http://%s.azurewebsites.net/%s",
-                    DockerRuntime.getInstance().getLatestWebAppName(), project.getName()));
-            props = DockerRuntime.getInstance().saveToProps(props);
-            ConfigFileUtil.saveConfig(project, props);
-        }
         return null;
     }
 
     private void buildImage(IProject project) throws Exception {
-        DockerClient dockerClient = DockerRuntime.getInstance().getDockerBuilder().build();
-        String destinationPath = project.getLocation() + Constant.DOCKER_CONTEXT_FOLDER + project.getName() + ".war";
-        WarUtil.export(project, destinationPath);
-        DockerUtil.buildImage(dockerClient, project, project.getLocation() + Constant.DOCKER_CONTEXT_FOLDER);
-        DockerRuntime.getInstance().setLatestArtifactName(project.getName());
+        if (project == null) {
+            throw new Exception("Project is null.");
+        }
+        String basePath = project.getLocation().toString();
+        DockerClient docker = DockerRuntime.getInstance().getDockerBuilder().build();
+        // export WAR file
+
+        String destinationPath = project.getLocation() + Constant.DOCKERFILE_FOLDER + project.getName() + ".war";
```

Comment:
the initial value is not used. So empty string is OK?

Reply:
yes, and I'd rather use null